### PR TITLE
65286_Authentication_User_CRUD

### PR DESCRIPTION
**Pull request by yasin-buet**
ffn1
ffn2

### CODE REVIEW

File: app/src/Template/Users/index.ctp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                <th class="actions"><?= __('Actions') ?></th>
+            </tr>
+        </thead>
+        <tbody>
+            <?php foreach ($users as $user): ?>
+            <tr>
+                <td><?= $this->Number->format($user->id) ?></td>
+                <td><?= h($user->username) ?></td>
+                <td><?= h($user->email) ?></td>
+                <td><?= h($user->role) ?></td>
```
Comment (hieubm) [inline app/src/Template/Users/index.ctp]: Why you have role here?
I think only admin can set role of users

File: app/src/Controller/AppController.php [hunk 2]
```diff
@@ -40,6 +40,18 @@ class AppController extends Controller
     public function initialize()
     {
         parent::initialize();
+        $this->loadComponent('Auth', [
+            'authorize' => ['Controller'],
+            'loginRedirect' => [
+                'controller' => 'News',
```
Comment (hieubm) [inline app/src/Controller/AppController.php]: We don't have News controller yet, so just redirect to home page

File: app/src/Template/Users/add.ctp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<div class="users form large-9 medium-8 columns content">
+    <?= $this->Form->create($user) ?>
+    <fieldset>
+        <legend><?= __('Add User') ?></legend>
+        <?php
+            echo $this->Form->input('username');
+            echo $this->Form->input('email');
+            echo $this->Form->input('password');
+            echo $this->Form->input('role');
```
Comment (hieubm) [inline app/src/Template/Users/add.ctp]: user can't register their role here

**Comment by asifbuetcse** [comment]
Ok.

File: app/src/Controller/AppController.php [hunk 4]
```diff
@@ -59,4 +71,17 @@ public function beforeRender(Event $event)
             $this->set('_serialize', true);
         }
     }
+
+    public function beforeFilter(Event $event)
+    {
+        $this->Auth->allow(['index', 'view', 'display']);
```
Comment (hieubm) [inline app/src/Controller/AppController.php]: You should remove it,
I think it's better to filter in each Controller, 
btw, if remove it, you don't have to check Auth in usercontroller/index

and user should not see all other users' detail

File: app/src/Controller/UsersController.php [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    /**
+     * Add method
+     *
+     * @return \Cake\Network\Response|void Redirects on successful add, renders view otherwise.
+     */
+    public function add()
+    {
+        $user = $this->Users->newEntity();
+        if ($this->request->is('post')) {
+              $this->request->data['role'] = 'user';
```
Comment (hieubm) [inline app/src/Controller/UsersController.php]: indent

File: app/src/Controller/UsersController.php [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+ * Users Controller
+ *
+ * @property \App\Model\Table\UsersTable $Users
+ */
+class UsersController extends AppController
+{
+    public function beforeFilter(Event $event)
+    {
+        parent::beforeFilter($event);
+        // $this->Auth->loginAction = array('controller' => 'News', 'action' => 'index');
```
Comment (hieubm) [inline app/src/Controller/UsersController.php]: remove comment

**Pull request merged**